rust/src/storage/file.rs

```diff
@@ -58,26 +63,14 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
-
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
-
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
-            .create_new(true)
-            .write(true)
-            .open(path)
-            .await
-            .map_err(|e| match e.kind() {
-                std::io::ErrorKind::AlreadyExists => StorageError::AlreadyExists(path.to_string()),
-                _ => StorageError::Io { source: e },
-            })?;
+        let tmp_file_name = format!("{}.temporary", Uuid::new_v4().to_string());
+        let tmp_path = self.join_path(&self.root, &tmp_file_name);
 
+        let mut f = fs::File::create(&tmp_path).await?;
```

Comment:
So I found this bug wile verifying a draft version of our S3 write design using TLA+. Although the chance of getting a UUID conflict is very very low, when it happens, we will corrupt the transaction log. We need to make sure the temporary files are created without overriding any existing files. This means we should reuse the original `OpenOptions::new().create_new(true)` logic to create the tmp file to guard against this unlikely event.

In the S3 write implementation, we will need to leverage dynamodb to guarantee uniqueness of temporary S3 keys as well.

Reply:
yeap, makes sense, will do

Reply:
@houqp did that, but also added a loop inside put_obj when creating tmp file, because if we propagate AlreadyExists back to the caller, then the ty_commit will call this function with incremented version, which actual is not that we want

Reply:
meaning that we address uuid conflict, not version number

Reply:
👍